SQL-639: PowerBi connection test

nbagnard opened the pull request.

pmeredit commented on odbc/src/api/definitions.rs:
```diff
@@ -111,131 +111,194 @@ pub enum UseBookmarks {
+    SQL_INDEX_KEYWORDS = 148,
+    SQL_INFO_SCHEMA_VIEWS = 149,
+    SQL_KEYSET_CURSOR_ATTRIBUTES1 = 150,
+    SQL_KEYSET_CURSOR_ATTRIBUTES2 = 151,
```
Also a good change.

pmeredit commented on odbc/src/api/functions.rs:
```diff
@@ -1840,448 +1840,425 @@ pub unsafe extern "C" fn SQLGetInfoW(
+                    const TIMEDATE_FUNCTIONS: u32 = SQL_FN_TD_CURRENT_TIMESTAMP
+                        | SQL_FN_TD_EXTRACT;
+                    i16_len::set_output_fixed_data(&TIMEDATE_FUNCTIONS, info_value_ptr, string_length_ptr)
+                }
```
[nit] We don't need these comments anymore after the enum name change :D

nbagnard replied:
You are right, I'll remove them!

nbagnard replied:
Done

pmeredit commented on odbc/src/handles/definitions.rs:
```diff
@@ -237,8 +237,8 @@ pub struct Statement {
 pub struct StatementAttributes {
-    pub app_row_desc: Pointer,
-    pub app_param_desc: Pointer,
+    pub app_row_desc: *mut MongoHandle,
```
this definitely makes this easier to work with.

pmeredit reviewed (approved):
LGTM other than the small nit :)

mattChiaravalloti commented on evergreen.yml:
```diff
@@ -202,7 +246,7 @@ functions:
           ./resources/run_adf.sh start &&
           cargo run --bin data_loader &&
-          cargo test integration
+          cargo test integration -- --nocapture
```
thank you 🙏

mattChiaravalloti commented on integration_test/tests/power_bi_tests.rs:
```diff
@@ -0,0 +1,231 @@
+                    str_len_ptr
+                )
+            );
+            println!(
```
Do we want to assert anything about these values (DBMS_NAME and VER) instead of just printing them? Or since they could vary (specifically version) are we ok just printing them and manually checking the values are correct?

nbagnard replied:
The idea is that we already have/will have unit tests which are validating the values so there is no need to re-check it again.
I print them out to make sure that using the built lib and the DM we get something out and that the buffer length and data match and that in case something looks wrong, we have more trace rather than less to go with when investigating.

mattChiaravalloti reviewed (approved):
lgtm 👍

nbagnard commented:
I've pushed the changes Oliver to run_adf.sh and evg.yml so that there is not more tasks hanging. The issue was that ADF started OK but Ubuntu 20.04 does not have 'netstat' so the check would fail, and then Evergreen would hang because ADF was still running in the background since we did not know it was, we were not stopping it. Oliver also added a task to copy the log so that we get more context.

The other change here is that I added functions to convert the sql return code to a string and get the diagnostic message when there is a failure and add that to the error the Integration test returns when something does not work as expected. This way instead of 
```
thread 'test_runner::integration_test' panicked at 'assertion failed: `(left == right)`
left: `Ok(())`,
right: `Err(OdbcFunctionFailed("sqltablesw", "SqlReturn(-1)"))`'
```
we get something a little more detailed like:
```
thread 'test_runner::integration_test' panicked at 'assertion failed: `(left == right)`
left: `Ok(())`,
right: `Err(OdbcFunctionFailed("sqltablesw", "ERROR", "[MongoDB][API] The feature SQLGetDiagFieldW is not implemented"))`', integration_test\src\test_runner\mod.rs:166:21

```
The changes are in `integration_test/tests/common/mod.rs` and used in `integration_test/src/test_runner/mod.rs` and `integration_test/tests/power_bi_tests.rs`.

mattChiaravalloti commented on integration_test/tests/common/mod.rs:
```diff
@@ -33,3 +34,45 @@ pub fn generate_default_connection_str() -> String {
     connection_string
 }
+
+#[allow(dead_code)]
```
```suggestion
#[test]
```
would `#[test]` be more appropriate than `#[allow(dead_code)]` since this could isn't "dead" but instead is used in codepaths guarded by `#[test]`? or maybe i'm losing the thread of how this function is being used!

nbagnard replied:
ah, this one was a mystery to me.
In lib.rs, this is flagged as code used for tests.
```
#[path = "../tests/common/mod.rs"]
#[cfg(test)]
mod common;
```
 It is also used similarly to generate_default_connection_str.
 So, I don't know what I missed, but it seemed more like a false positive that Clippy was flagging so I put the tag to allow dead code.
 If I missed something and I can clean this up, let me know.
 #[test] does not work because this is not a test, but a helper function used by multiple tests.

mattChiaravalloti replied:
Oh right right, I meant `#[cfg(test)]` 🤦‍♀️  But if the entire module is already marked with that, then I'm not sure adding it at the function level will do anything anyway!

mattChiaravalloti commented on integration_test/tests/common/mod.rs:
```diff
@@ -33,3 +34,45 @@ pub fn generate_default_connection_str() -> String {
+    }
+}
+
+#[allow(dead_code)]
```
same question/answer would apply here! ignore if `#[test]` is not correct.

mattChiaravalloti reviewed (approved):
The run_adf changes look good 👍

pmeredit reviewed (approved):
LGTM!
